Sources/NIOHTTP1/HTTPUpgradeHandler.swift

```diff
@@ -191,48 +203,45 @@ public class HTTPServerUpgradeHandler: ChannelInboundHandler {
                 ctx.fireErrorCaught(error)
                 continue
             }
-
-            // We are now upgrading, any further data should be buffered and replayed.
-            self.upgrading = true
-
-            // Before we finish the upgrade we have to remove the HTTPDecoder and any other non-Encoder HTTP
-            // handlers from the pipeline, to prevent them parsing any more data. We'll buffer the data until
-            // that completes.
-            // While there are a lot of Futures involved here it's quite possible that all of this code will
-            // actually complete synchronously: we just want to program for the possibility that it won't.
-            // Once that's done, we send the upgrade response, then remove the HTTP encoder, then call the
-            // internal handler, then call the user code, and then finally when the user code is done we do
-            // our final cleanup steps, namely we replay the received data we buffered in the meantime and
-            // then remove ourselves from the pipeline.
-            _ = self.removeExtraHandlers(ctx: ctx).then {
-                self.sendUpgradeResponse(ctx: ctx, upgradeRequest: request, responseHeaders: responseHeaders)
-            }.then {
-                self.removeHandler(ctx: ctx, handler: self.httpEncoder)
-            }.map { (_: Bool) in
-                self.upgradeCompletionHandler(ctx)
-            }.then {
-                upgrader.upgrade(ctx: ctx, upgradeRequest: request)
-            }.map {
-                ctx.fireUserInboundEventTriggered(HTTPUpgradeEvents.upgradeComplete(toProtocol: proto, upgradeRequest: request))
-
-                self.upgrading = false
-
-                // We unbuffer any buffered data here and, if we sent any,
-                // we also fire readComplete.
-                let bufferedMessages = self.receivedMessages
-                self.receivedMessages = []
-                bufferedMessages.forEach { ctx.fireChannelRead($0) }
-                if bufferedMessages.count > 0 {
-                    ctx.fireChannelReadComplete()
+           
+            return {
```

Comment:
There's no need to heap allocate this, closing over the `ChannelHandlerContext`. Instead, factor this out to a separate function and just use a state variable.

Reply:
Eh, I guess that's fine? It feels super weird to me, but ok. We can't use `@convention(thin)` because as @normanmaurer notes this actually closes over a *bunch* of state.

We could encapsulate this in a struct instead, to avoid the heap allocation.

Reply:
I am happy with the struct as well I just thought this one is a bit nicer and it should not matter. That said I am fine either way, just tell me what to do :)

Reply:
For now we can allocate the closure, we can always remove it later.